xoto3/dynamodb/write_versioned/api2.py

```diff
@@ -0,0 +1,132 @@
+"""
+A table and type-oriented API for using versioned_transact_write_items.
+
+All of these things are simple bolt-ons for the underlying
+functionality, and may safely be ignored.
+"""
+from copy import deepcopy
+from typing import Callable, Generic, Optional, TypeVar, Union
+
+from .modify import define_table, delete, presume, put
+from .read import get, require
+from .types import Item, ItemKey, TransactionBuilder, VersionedTransaction
+
+T = TypeVar("T")
+
+
+class TypedTable(Generic[T]):
+    """In practice, it's nice to have table-aware closures of these
+    functions when dealing with a table, versus providing the string
+    table name to every call. This is especially valuable if you're
+    dealing with only a single table over and over again, and there's
+    some kind of ser/de you want to run consistently against that
+    table.
+
+    Note that your type_deserializer is now the expected method of
+    performing the standard deepcopy-on-get to prevent mutating the
+    immutable items retrieved from a VersionedTransaction.  Therefore
+    the standard copy-on-get will not be performed. If you have no
+    need for additional type deserialization, you should provide
+    `copy.deepcopy` to retain the standard, safe behavior.
+    """
+
+    def __init__(
+        self,
+        table_name: Union[Callable[[], str], str],
+        type_deserializer: Callable[[Item], T],
+        # from the database to your code - on the read path
+        type_serializer: Callable[[T], Item],
+        # from your code to the database - on the write path
+        type_name: str = "Item",
+    ):
+        if isinstance(table_name, str):
+            self.lazy_table_name = lambda: table_name
+        else:
+            self.lazy_table_name = table_name
+        self.type_name = type_name
+        self.type_deserializer = type_deserializer
+        self.type_serializer = type_serializer
+
+    def get(self, key: ItemKey) -> Callable[[VersionedTransaction], Optional[T]]:
+        """Get an item from the database if it exists"""
+
+        def deser_opt(item: Optional[Item]) -> Optional[T]:
+            return self.type_deserializer(item) if item else None
+
+        return lambda vt: deser_opt(
+            get(vt, self.lazy_table_name(), key, copy=False, nicename=self.type_name)
```

Comment:
duh. You're still calling the original functions. These are just convenience wrappers. Like you said. Got it.